glib-macros/src/clone.rs

```diff
@@ -1,952 +1,487 @@
-        Some(TokenTree::Ident(i)) if i.to_string() == "move" => false,
-        Some(TokenTree::Ident(i)) if i.to_string() == "async" => true,
-        Some(TokenTree::Ident(i)) if i.to_string() == "default" => {
-            let span = parts.next_span;
-            let ret = return_kind(parts)?;
-            let err = format!("Missing `@` before `{}`", ret.keyword());
-            return Err(parts.generate_error_with_span(&err, span));
-        }
-        Some(TokenTree::Punct(p)) if p.to_string() == "|" => {
-            return Err(parts.generate_error_with_next_span(
-                "Closure needs to be \"moved\" so please add `move` before closure",
-            ));
-        }
-        _ => {
-            return Err(
-                parts.generate_error_with_next_span("Missing `move` and closure declaration")
+        let outer_before = self
+            .captures
+            .iter()
+            .map(|c| c.outer_before_tokens(&crate_ident));
+        let inner_before = self.captures.iter().map(|c| {
+            c.inner_before_tokens(
+                &crate_ident,
+                &self.default_return,
+                &default_return_closure_ident,
             )
-        }
-    };
-    parts.next();
-    if is_async {
-        return check_async_syntax(parts);
-    }
-    match parts.peek() {
-        Some(TokenTree::Punct(p)) if p.to_string() == "|" => {}
-        Some(x) => {
-            let err = format!("Expected closure, found `{x}`");
-            return Err(parts.generate_error_with_next_span(&err));
-        }
-        None => return Err(parts.generate_error("Expected closure")),
-    }
-    parts.next();
-
-    let closure = closure(parts)?;
-    match parts.peek() {
-        Some(TokenTree::Ident(i)) if i.to_string() == "async" => {
-            parts.next();
-            check_move_after_async(parts)?;
-            match parts.peek() {
-                Some(TokenTree::Group(g)) if g.delimiter() == Delimiter::Brace => {
-                    Ok(BlockKind::ClosureWrappingAsync(closure))
-                }
-                // The next matchings are for better error messages.
-                Some(TokenTree::Punct(p)) => {
-                    let err = format!("Expected block after `| async move`, found `{p}`");
-                    Err(parts.generate_error_with_next_span(&err))
-                }
-                Some(TokenTree::Group(g)) => {
-                    let err = format!(
-                        "Expected block after `| async move`, found `{}`",
-                        delimiter_to_string(g.delimiter(), true),
-                    );
-                    Err(parts.generate_error_with_next_span(&err))
+        });
+
+        let default_return_closure =
+            if let DefaultReturn::Return(Some(ref closure)) = self.default_return {
+                Some(quote! {
+                    let #default_return_closure_ident = #closure;
+                })
+            } else {
+                None
+            };
```

Comment:
I think a `match` would look a bit better:

```suggestion
        let default_return_closure = match self.default_return {
             DefaultReturn::Return(Some(ref closure)) => Some(quote! {
                 let #default_return_closure_ident = #closure;
            }),
            _ => None,
        };
```